yorkie/auth/webhook.go

```diff
@@ -49,29 +54,70 @@ func VerifyAccess(ctx context.Context, be *backend.Backend, info *types.AccessIn
 		return err
 	}
 
-	// TODO(hackerwins): We need to apply retryBackoff in case of failure
-	resp, err := http.Post(
-		be.Config.AuthorizationWebhookURL,
-		"application/json",
-		bytes.NewBuffer(reqBody),
-	)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		if err := resp.Body.Close(); err != nil {
-			log.Logger.Error(err)
+	return withExponentialBackoff(be.Config, func() (int, error) {
+		resp, err := http.Post(
+			be.Config.AuthorizationWebhookURL,
+			"application/json",
+			bytes.NewBuffer(reqBody),
+		)
+		if err != nil {
+			return 0, err
 		}
-	}()
 
-	authResp, err := types.NewAuthWebhookResponse(resp.Body)
-	if err != nil {
+		defer func() {
+			if err := resp.Body.Close(); err != nil {
+				log.Logger.Error(err)
+			}
+		}()
+
+		if http.StatusOK != resp.StatusCode {
+			return resp.StatusCode, fmt.Errorf("%d: %w", resp.StatusCode, ErrUnexpectedStatusCode)
+		}
+
+		authResp, err := types.NewAuthWebhookResponse(resp.Body)
+		if err != nil {
+			return resp.StatusCode, err
+		}
+
+		if !authResp.Allowed {
+			return resp.StatusCode, fmt.Errorf("%s: %w", authResp.Reason, ErrNotAllowed)
+		}
+
+		return resp.StatusCode, nil
+	})
+}
+
+func withExponentialBackoff(config *backend.Config, webhookFn func() (int, error)) error {
+	statusCode, err := webhookFn()
+	if !shouldRetry(statusCode) {
 		return err
 	}
 
-	if !authResp.Allowed {
-		return fmt.Errorf("%s: %w", authResp.Reason, ErrNotAllowed)
+	var retries uint64
+	for {
+		if retries == config.AuthorizationWebhookMaxRetries {
+			return err
+		}
+
+		time.Sleep(waitInterval(retries, config.AuthorizationWebhookMaxWaitInterval))
+
+		statusCode, err := webhookFn()
+		if !shouldRetry(statusCode) {
+			return err
+		}
+		retries++
```

Comment:
1. It would be nice to be able to cancel with the context passed from the caller.
2. I think It would be good to explicitly return ErrWaitTimeout.

```suggestion
	var retries uint64
	for retries < config.AuthorizationWebhookMaxRetries {
		waitBeforeRetry := waitInterval(retries, config.AuthorizationWebhookMaxWaitIntervalSec)
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(waitBeforeRetry):
		}

		statusCode, err := webhookFn()
		if !shouldRetry(statusCode) {
			return err
		}
		retries++
	}
	
	return ErrWaitTimeout
```

Reply:
1. I agree. I think what you said is related to the opinion you gave in point 1 of my discussion. What do you think about adding the Timeout option of the API to the Agent?

2. I thought it would be easier for users to figure out the problem by returning the last state of the webhook rather than a Timeout error after the last retry. Could you explain a bit more about the benefit of returning a timeout error on failure after the last retry?